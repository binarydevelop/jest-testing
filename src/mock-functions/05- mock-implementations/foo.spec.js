describe('Mock Implementations', () => {
    test('mockingImplementation', () => {
        const myMockFn = jest.fn();

        myMockFn((val) => console.log(val));    
    })

    test('mockingImplementationOnce', () => {
        const myMockFn = jest
        .fn(() => 'default')
        .mockImplementationOnce(() => 'first call')
        .mockImplementationOnce(() => 'second call');
      
      console.log(myMockFn(), myMockFn(), myMockFn(), myMockFn());
      // > 'first call', 'second call', 'default', 'default'        
    })

    test('mocking This', () => {
        const myMockFn = {
            thisFunction : jest.fn().mockReturnThis() // it will return this object 
        } 
        myMockFn.thisFunction();
        expect(myMockFn.thisFunction).toHaveBeenCalled();
        
    })

    test('mock Names', () => {
        const myMockFn = jest
        .fn()
        .mockReturnValue('default')
        .mockImplementation(scalar => 42 + scalar)
        .mockName('add');

        // expect(myMockFn).toBeCalled(); 
        // will throw error with the name provided
    })

    test('Custom Matchers', () => {
        const mockFunction = jest.fn().mockReturnValue("Mocked")
        const arg1 = "Mocking"
        const arg2 = "Custom Matcher"

        mockFunction(arg1, arg2);

        expect(mockFunction).toBeCalled();
        expect(mockFunction).toHaveBeenCalledWith(arg1, arg2);
        expect(mockFunction).toHaveBeenLastCalledWith(arg1, arg2);
        expect(mockFunction).toMatchSnapshot();
    })

})
