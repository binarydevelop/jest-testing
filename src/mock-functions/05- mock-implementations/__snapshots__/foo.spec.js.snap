// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mock Implementations Custom Matchers 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Mocking",
      "Custom Matcher",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "Mocked",
    },
  ],
}
`;
